### PR TITLE
[TO_REVIEW] Modification source_target_merge function behaviours

### PR DESCRIPTION
**Pull request by YanisLalou**
Issue: #61 
Change the name of `source_target_merge`
Allow an N dimension array as an input for `source_target_merge`
Add test cases

For `source_target_merge`:
- Sample domain input should be optional
- Y_target can be equal to None (check there's only one None given in all pairs)
- X_source/target can be list of arrays, check all sizes are coherent

### CODE REVIEW

File: skada/utils.py [hunk 1]
```diff
@@ -243,3 +243,56 @@ def source_target_split(
...
+    if X_source.shape[0] > 0 and X_target.shape[0] > 0:
+        output = np.zeros_like(
+            np.concatenate((X_source, X_target)), dtype=X_source.dtype
+        )
+        output[sample_domain >= 0] = X_source
+        output[sample_domain < 0] = X_target
+
+    elif X_source.shape[0] > 0:
+        output = np.zeros_like(
+            X_source, dtype=X_target.dtype
```
Comment (tgnassou) [inline skada/utils.py:288]: `dtype=X_source.dtype`

Comment (tgnassou) [reply skada/utils.py:288]: For `X_target` a empty set I don't know what will be the `dtype`

**Review by tgnassou — APPROVED** [review]

File: skada/utils.py [hunk 2]
```diff
@@ -243,3 +243,56 @@ def source_target_split(
     return list(chain.from_iterable(
         (a[source_idx], a[~source_idx]) for a in arrays
     ))
+
+
+def source_target_merge(
```
Comment (kachayev) [inline skada/utils.py:248]: I think we should match split API: accept as many pairs of 'things' as necessary, alongside with a named argument 'sample_domain'. 

So I can do something like

```python 
X, y = source_target_merge(X_source, X_target, y_source, y_target, sample_domain=sample_domain)
```

**Review by kachayev — CHANGES_REQUESTED** [review]

File: skada/utils.py (hunk 2 above)
Comment (rflamary) [reply skada/utils.py:248]: Do you mean : 

```python
X, y, sample_domain = source_target_merge(X_source, X_target, y_source, y_target)
```

insterad?

Comment (YanisLalou) [reply skada/utils.py:248]: In `source_target_split()` we don't output back the sample domain, just the splitted arrays. Makes sense to do the same here

Comment (kachayev) [reply skada/utils.py:248]: Right, as we have this function to pair up with the split functionality, we shouldn't encourage swapping sample_domain to a new one, even though it's technically possible (it would mean that we have to return it with every operation). Moreover, it might easily lead to 'accidental' loss of the information when having multiple sources/targets.

**Comment by YanisLalou** [comment]
For now `source_target_merge()` accepts an even number of arrays as input + the sample domain array.
It works by considering the arrays by duo (the first one is considered as the source and the second the target).
Thus I'm wondering:
+ if it would be better to accept the duos as tuples rather than directly in the *arrays.
+ Don't know if the doc is clear enough 

Would love your feedbacks

**Comment by kachayev** [comment]
Regarding pairs, you are right, it might be better with pairs. But it's common to have pairs of ndarrays as an input type... So it might also be a bit confusing. Let's keep the current implementation to see how it works in practice. We will be able to change it if necessary.

**Comment by kachayev** [comment]
@YanisLalou is this one ready for review?

**Comment by YanisLalou** [comment]
> @YanisLalou is this one ready for review?

Almost, there's one codevoc test that I still need to implement + check if I've managed to handle all possible cases.
But otherwise yes you can start reviewing it + if you find any edge cases that I didn't foreseen, that would really help me

File: skada/tests/test_utils.py [hunk 3]
```diff
@@ -290,3 +294,126 @@ def test_extract_source_indices():
...
+        pos_source=0.1,
+        pos_target=0.9,
+        n_samples_source=n_samples_source,
+        n_samples_target=n_samples_target,
+        random_state=0,
+        return_X_y=True,
+    )
+
+    X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+    y_source, y_target = source_target_split(y, sample_domain=sample_domain)
```
Comment (kachayev) [inline skada/tests/test_utils.py:312]: Shouldn't `source_target_split` function accept both `X, y` together?

File: skada/tests/test_utils.py [hunk 4]
```diff
@@ -290,3 +294,126 @@ def test_extract_source_indices():
...
+    # Test y_target = None + Consistent number of samples in source-target
+    _ = source_target_merge(
+        X_source,
+        X_target,
+        y_source,
+        None,
+        sample_domain=sample_domain
+    )
+
+    # Test 2 None in a duo arrays
```
Comment (kachayev) [inline skada/tests/test_utils.py:402]: 'array pairs' or 'pairs of arrays'

File: skada/utils.py [hunk 5]
```diff
@@ -247,3 +251,223 @@ def source_target_split(
...
+    sample_domain : Optional[np.ndarray] = None
+) -> np.ndarray:
+    f""" Merge source and target domain data based on sample domain labels.
+
+    Parameters
+    ----------
+    *arrays : sequence of array-like
+        (, n_features). The number of arrays must be even
+        since we consider duos of source-target arrays each time.
+        Each duo should have at least one non-empty array.
```
Comment (kachayev) [inline skada/utils.py:267]: Each pair

File: skada/utils.py [hunk 6]
```diff
@@ -247,3 +251,223 @@ def source_target_split(
...
+    # ################ Check sample_domain #################
+    # If sample_domain is None, we need to infer it from the target array
+    if sample_domain is None or sample_domain.shape[0] == 0:
+        # We need to infer the domain from the target array
+        warnings.warn(
+            "sample_domain is None or empty, it will be inferred from the arrays"
+        )
+
+        # By assuming that the first array is the source and the second the target
+        index_something_source = 0
```
Comment (kachayev) [inline skada/utils.py:357]: Why `something`?

File: skada/utils.py [hunk 7]
```diff
@@ -247,3 +251,223 @@ def source_target_split(
...
+            y_type = type_of_target(arrays[duo_index])
+            if y_type == 'binary' or y_type == 'multiclass':
+                arrays[index_is_empty] = (
+                    _DEFAULT_MASKED_TARGET_CLASSIFICATION_LABEL *
+                    np.ones(
+                        (sample_domain.shape[0] - arrays[duo_index].shape[0],) +
+                        arrays[duo_index].shape[1:]
+                    )
+                )
+            else:
```
Comment (kachayev) [inline skada/utils.py:404]: The only difference between it and else branches is the value of the label. Thus the code could be simplified.

**Review by kachayev — CHANGES_REQUESTED** [review]
Overall comment, let's replace 'duo' with 'pair'. I mentioned it multiple times in different places in the code but certainly not everywhere.

File: skada/utils.py [hunk 8]
```diff
@@ -247,3 +253,219 @@ def source_target_split(
         (a[source_idx], a[~source_idx]) if a is not None else (None, None)
         for a in arrays
     ))
+
+
+def source_target_merge(
+    *arrays,
+    sample_domain : Optional[np.ndarray] = None
+) -> np.ndarray:
```
Comment (kachayev) [inline skada/utils.py:261]: Missed this one last time: the return type should be a tuple of ndarrays, moreover, of unknown size. `Sequence` would be probably the closest container type for such situations.

**Review by kachayev — CHANGES_REQUESTED** [review]
